release-19.1: sql: fix the auto-retry counter in stats + log it in statement/audit logs

knz opened the pull request:
Backport 1/2 commits from #38028.
Backport 1/1 commits from #38051.
Backport 1/1 commits from #38078.

/cc @cockroachdb/release

---

knz committed:
sql: fix the retry counter in the statement statistics

The statement statistics contain a column `max_retries` intended
to show the maximum number of times a statement was retried.

Due to a bug, this counter was previously always zero, *even for
statements that were actually retried*. We hadn't noticed because
there was no test to ensure that the value would increase.

This patch fixes the bugs and verifies that the value does indeed
increase.

Release note (bug fix): The `max_retries` column in the statement
statistics now properly displays the maximum number of times
a statement was retried.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/38035)
<!-- Reviewable:end -->

bdarnell commented:
I don't think we should backport the second commit. Changing the exec/audit log format in a patch release is a bad idea. (Adding a column at the end is backwards-compatible if you've designed your parser to accept unknown extra columns, but not everyone does that right and it's not something we want to risk in a patch release).

knz committed:
sql: deflake TestLogic//crdb_internal/max_retry_counter

The retry timeout was too short for `make stress`.
Instead this patch copies the timeout already used in the other
logic test file `txn`. I verified manually that `make stress`
is pleased with the new timeout.

Also fix a typo in the test comment.

Release note: None

knz commented:
Ok I removed the 2nd commit.

knz committed:
sql: logic_test/crdb_internal: add missing comments

Requested by @rytaft.

Release note: None

knz commented:
@jordanlewis friendly ping

jordanlewis commented:
LGTM, merge away. Thanks @knz!
